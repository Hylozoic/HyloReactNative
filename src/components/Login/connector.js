import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { login, loginWithFacebook, loginWithGoogle } from './actions'

function mapStateToProps (state) {
  return {
    error: state.session.loginError,
    defaultEmail: state.session.defaultLoginEmail
  }
}

function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators({
      login,
      loginWithFacebook,
      loginWithGoogle
    }, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
