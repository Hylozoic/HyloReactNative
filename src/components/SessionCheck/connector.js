import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { checkSession } from './actions'
import getMe from '../../store/selectors/getMe'
import { get } from 'lodash/fp'

export function mapStateToProps (state) {
  const signupInProgress = get('settings.signupInProgress', getMe(state))

  return {
    loggedIn: state.session.loggedIn && !signupInProgress
  }
}

export function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators({
      checkSession
    }, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
