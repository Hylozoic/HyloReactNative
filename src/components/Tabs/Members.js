import React from 'react'

import WelcomeScene from '../WelcomeScene'
import Header from './Header'

export default class Members extends React.Component {
  static navigationOptions = ({navigation}) => (Header(navigation))
  render () {
    const { navigation } = this.props
    return <WelcomeScene navigation={navigation} />
  }
}
