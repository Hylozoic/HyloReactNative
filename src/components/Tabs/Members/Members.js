import React from 'react'
import { View, FlatList, Text, TouchableOpacity, TextInput, Image } from 'react-native';
import Header from '../Header'
import Avatar from '../../Avatar'
import PopupMenuButton from '../../PopupMenuButton'
import Icon from '../../Icon'
import { DEFAULT_BANNER } from '../../../store/models/Community'
import styles from './Members.styles'
import { some, values, keys, isUndefined, isEmpty, debounce } from 'lodash/fp'
import { focus } from '../../../util/textInput'
const title = 'Members'

export default class Members extends React.Component {
  static navigationOptions = ({navigation}) => (Header(navigation, title))

  fetchOrShowCached () {
    const { hasMore, members, fetchMembers } = this.props
    if (isEmpty(members) && hasMore !== false) fetchMembers()
  }

  componentDidMount () {
    this.fetchOrShowCached()
  }

  componentDidUpdate (prevProps) {
    if (!prevProps) return

    if (some(key => this.props[key] !== prevProps[key], [
        'slug',
        'networkSlug',
        'sortBy',
        'search',
      ])) {
      this.fetchOrShowCached()
    }
  }

  render () {
    const { community, subject, sortBy, setSort, fetchMoreMembers } = this.props

    const sortKeys = sortKeysFactory(subject)

    const onSearch = debounce(300, text => {
      this.props.setSearch(text)
    })

    const onSelect = (action, index) => {
      if (!isUndefined(index)) {
        setSort(keys(sortKeys)[index])
      }
    }

    let { members } = this.props

    // Members need to be even since it's rows of 2
    if (members.length % 2 > 0) {
      members.push({id: -1})
    }

    const headerComponent = <View>
      <Banner community={community} all={!community} />
      <View style={styles.listControls}>
        <View style={styles.searchWrapper}>
          <Icon style={styles.searchIcon} name='Search' size={30} onPress={() => focus(this.searchRef)}/>
          <TextInput placeholder='Search Members'
                     ref={ref => this.searchRef = ref}
                     onChangeText={onSearch}
                     underlineColorAndroid='transparent' style={styles.searchInput} />
        </View>

        <PopupMenuButton actions={values(sortKeys)} onSelect={onSelect}>
          <View style={styles.sortBy}>
            <Text style={styles.sortByText}>{sortKeys[sortBy]}</Text>
            <Icon name='ArrowDown' style={styles.downArrow} />
          </View>
        </PopupMenuButton>
      </View>
    </View>

    return <View style={styles.container}>
      <FlatList
        data={members}
        numColumns='2'
        renderItem={({item}) => {
          if (item.name) {
            return <Member member={item} showMember={this.props.showMember}/>
          } else {
            return <View style={styles.cell} />
          }
        }}
        onEndReached={fetchMoreMembers}
        keyExtractor={(item, index) => item.id}
        ListHeaderComponent={headerComponent}
      />
    </View>
  }
}

function Member ({ member, showMember }) {
  return <TouchableOpacity onPress={() => showMember(member.id)}
                           style={[styles.cell, styles.memberCell]} >
    <View style={styles.avatarSpacing}>
      <Avatar avatarUrl={member.avatarUrl} dimension={72} />
    </View>
    <Text style={styles.memberName}>{member.name}</Text>
    {member.location && <Text style={styles.memberLocation}>{member.location}</Text>}
    <Text style={styles.memberBio}>{member.bio}</Text>
  </TouchableOpacity>
}

function Banner ({ community, all }) {
  let bannerUrl, name

  if (all) {
    name = 'All Communities'
    bannerUrl = DEFAULT_BANNER
  } else if (!community) {
    return null
  } else {
    ({ bannerUrl, name } = community)
  }

  return <View style={styles.bannerContainer}>
    <Image source={{uri: bannerUrl}} style={styles.image} />
    <View style={styles.titleRow}>
      <Text style={styles.name}>{name}</Text>
    </View>
  </View>
}

// these keys must match the values that hylo-node can handle
function sortKeysFactory (subject) {
  const sortKeys = {
    name: 'Name',
    location: 'Location'
  }
  if (subject !== 'network') sortKeys['join'] = 'Newest'
  return sortKeys
}
