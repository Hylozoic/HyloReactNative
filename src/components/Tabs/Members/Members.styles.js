import { capeCod } from '../../../style/colors'

export default {
  container: {
    backgroundColor: '#FFF'
  },
  bannerContainer: {
    zIndex: 10
  },
  image: {
    height: 100
  },
  titleRow: {
    position: 'absolute',
    left: 0,
    right: 0,
    flexDirection: 'row',
    marginLeft: 16,
    marginRight: 16,
    top: 60
  },
  name: {
    fontSize: 24,
    fontWeight: 'bold',
    color: 'white',
    backgroundColor: 'transparent'
  },
  cell: {
    flex: 1,
    margin: 8,
    height: 100,
    alignItems: 'center'
  },
  memberCell: {
    backgroundColor: '#FFF',
    borderColor: '#EAEBEB',
    borderStyle: 'solid',
    borderWidth: 1,
    borderRadius: 4
  },
  listControls: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 10,
    marginRight: 16,
    marginLeft: 16
  },
  listControl: {
    flexDirection: 'row'
  },
  optionText: {
    fontSize: 12,
    color: capeCod
  },
  downArrow: {
    top: 2,
    marginLeft: 4
  },
}
