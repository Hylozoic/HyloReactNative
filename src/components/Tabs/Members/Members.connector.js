import { connect } from 'react-redux'
import { omit, get } from 'lodash/fp'

import { ALL_COMMUNITIES_ID } from '../../../store/models/Community'
import {
  FETCH_MEMBERS,
  getHasMoreMembers,
  fetchMembers,
  getSort,
  setSort,
  getSearch,
  setSearch,
  getMembers
} from './Members.store'
import getMe from '../../../store/selectors/getMe'
import getCommunity from '../../../store/selectors/getCommunity'
import { mapWhenFocused, mergeWhenFocused } from 'util/connector'

function makeFetchOpts (props) {
  const { community } = props

  return {
    ...omit('community', props),
    subject: community ? 'community' : 'all-communities',
    slug: get('slug', community) || ALL_COMMUNITIES_ID
  }
}

export function mapStateToProps (state, props) {
  const currentUser = getMe(state, props)
  const communityId = state.currentCommunity ||
    (currentUser && get('id', currentUser.lastViewedCommunity()))
  const community = getCommunity(state, {id: communityId})
  const slug = community && community.slug

  const search = getSearch(state)
  const sortBy = getSort(state)

  return {
    currentUser,
    communityId,
    community,
    hasMore: getHasMoreMembers(state, {slug, search, sortBy}),
    members: getMembers(state, {slug, search, sortBy}),
    pending: state.pending[FETCH_MEMBERS],
    search,
    slug,
    sortBy
  }
}

export function mapDispatchToProps (dispatch, { navigation }) {
  return {
    setSort: sort => dispatch(setSort(sort)),
    setSearch: search => dispatch(setSearch(search)),
    fetchMembers: opts => dispatch(fetchMembers(opts)),
    showMember: id => navigation.navigate('MemberProfile', {id}),
    updateBadges: badgeFlags => navigation.setParams(badgeFlags)
  }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { sortBy, hasMore, pending, search, members, community } = stateProps
  const fetchOpts = makeFetchOpts({community, sortBy, search})
  const fetchMembers = () =>
    !!community && dispatchProps.fetchMembers(fetchOpts)
  const offset = members.length
  const fetchMoreMembers = hasMore && !pending
    ? () => dispatchProps.fetchMembers({ ...fetchOpts, offset })
    : () => {}

  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    fetchMoreMembers,
    fetchMembers
  }
}

export default connect(
  mapWhenFocused(mapStateToProps),
  mapDispatchToProps,
  mergeWhenFocused(mergeProps)
)
