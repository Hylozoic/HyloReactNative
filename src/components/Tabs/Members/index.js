import { withNavigationFocus } from 'react-navigation-is-focused-hoc'

import component from './Members'
import connector from './Members.connector'

export default withNavigationFocus(connector(component))
