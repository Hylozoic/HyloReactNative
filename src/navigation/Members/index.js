import component from './Members'
import connector from './Members.connector'
import { withNavigationFocus } from '@react-navigation/compat'

export default withNavigationFocus(connector(component))
