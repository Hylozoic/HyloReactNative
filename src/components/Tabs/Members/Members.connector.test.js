import { mapStateToProps, mapDispatchToProps, mergeProps, makeFetchOpts } from './Members.connector'
import { MODULE_NAME } from './Members.store'
import orm from '../../../store/models'

describe('makeFetchOpts', () => {
  it('handles a network', () => {
    const props = {
      network: {
        slug: 'netslug'
      },
      community: null,
      sortBy: 'join',
      search: 'fee'
    }
    expect(makeFetchOpts(props)).toMatchSnapshot()
  })

  it('handles a community', () => {
    const props = {
      network: null,
      community: {
        slug: 'comslug'
      },
      sortBy: 'join',
      search: 'fee'
    }
    expect(makeFetchOpts(props)).toMatchSnapshot()
  })

  it('handles no community or network', () => {
    const props = {
      network: null,
      community: null,
      sortBy: 'join',
      search: 'fee'
    }
    expect(makeFetchOpts(props)).toMatchSnapshot()
  })
})

describe('mapStateToProps', () => {
  it('handles null value for lastViewedCommunity', () => {
    const session = orm.session(orm.getEmptyState())
    const community = session.Community.create({id: 10, slug: 'tom'})
    session.Me.create({
      id: 123,
      memberships: [session.Membership.create({
        id: '345',
        community: community.id,
        hasModeratorRole: true
      })]
    })

    const state = {
      orm: session.state,
      [MODULE_NAME]: {},
      pending: {},
      queryResults: {}
    }
    session.Community.create({})
    expect(mapStateToProps(state)).toMatchSnapshot()
  })
})

describe('mapDispatchToProps', () => {
  it('returns the right keys', () => {
    expect(mapDispatchToProps(() => {})).toMatchSnapshot()
  })
})

describe('mergeProps', () => {
  let dispatchProps, stateProps

  beforeEach(() => {
    dispatchProps = { fetchMembers: jest.fn() }
    stateProps = { hasMore: true, members: [], pending: false }
  })

  it('makes fetchMembers a no-op when there is no community or network', () => {
    const props = mergeProps(stateProps, dispatchProps)
    expect(props.fetchMembers()).toEqual(undefined)
    expect(dispatchProps.fetchMembers).not.toBeCalled()
  })

  it("makes fetchMoreMembers an empty function if there ain't no more", () => {
    stateProps.hasMore = false
    const actual = mergeProps(stateProps, dispatchProps)
    actual.fetchMoreMembers()
    expect(dispatchProps.fetchMembers).not.toHaveBeenCalled()
  })

  it('calls fetchMembers with the correct offset when fetchMoreMembers is called', () => {
    stateProps.members = [ 'so', 'many', 'members' ]
    const actual = mergeProps(stateProps, dispatchProps)
    actual.fetchMoreMembers()
    expect(dispatchProps.fetchMembers.mock.calls[0][0].offset).toBe(stateProps.members.length)
  })
})
