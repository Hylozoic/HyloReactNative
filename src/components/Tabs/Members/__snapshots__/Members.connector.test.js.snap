// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeFetchOpts handles a community 1`] = `
Object {
  "search": "fee",
  "slug": "comslug",
  "sortBy": "join",
  "subject": "community",
}
`;

exports[`makeFetchOpts handles a network 1`] = `
Object {
  "search": "fee",
  "slug": "netslug",
  "sortBy": "name",
  "subject": "network",
}
`;

exports[`makeFetchOpts handles no community or network 1`] = `
Object {
  "search": "fee",
  "slug": "all-communities",
  "sortBy": "join",
  "subject": "all-communities",
}
`;

exports[`mapStateToProps handles null value for lastViewedCommunity 1`] = `
Object {
  "community": null,
  "currentUser": SessionBoundModel {
    "_fields": Object {
      "id": 123,
      "memberships": undefined,
      "messageThreads": undefined,
      "notifications": undefined,
      "posts": undefined,
      "skills": undefined,
    },
    "id": 123,
  },
  "fetchOpts": Object {
    "search": undefined,
    "slug": "all-communities",
    "sortBy": undefined,
    "subject": "all-communities",
  },
  "hasMore": undefined,
  "isAll": true,
  "members": Array [],
  "network": null,
  "pending": undefined,
  "search": undefined,
  "slug": "all-communities",
  "sortBy": undefined,
  "subject": "all-communities",
}
`;
